packages/libs/client/src/v2/pact.ts

```diff
@@ -0,0 +1,238 @@
+    Partial<ICommand> | ((cmd: Partial<ICommand>) => Partial<ICommand>)
+  > = [first, ...rest] as any;
+  const command = args.reduce<Partial<ICommand>>((acc, item) => {
+    const part = typeof item === 'function' ? item(acc) : item;
+    if (part.payload !== undefined) {
+      if (acc.payload !== undefined) {
+        throw Error('Only one payload is acceptable');
+      }
+      acc.payload = part.payload;
+    }
+    if (part.meta !== undefined) {
+      acc.meta = { ...acc.meta, ...part.meta };
+    }
+    if (part.nonce !== undefined) {
+      acc.nonce = part.nonce;
+    }
+    if (part.networkId !== undefined) {
+      acc.networkId = part.networkId;
+    }
+    if (part.signers !== undefined) {
+      part.signers.forEach((signer) => {
+        acc.signers ??= [];
+        const prev = acc.signers.find(({ pubKey }) => signer.pubKey === pubKey);
+        if (prev !== undefined) {
+          prev.clist = [...(prev.clist ?? []), ...(signer.clist ?? [])];
+        } else {
+          acc.signers.push(signer);
+        }
+      });
+    }
+    return acc;
+  }, {} as Partial<ICommand>);
+  const dateInMs = Date.now();
+  command.nonce = command.nonce ?? `kjs-${dateInMs}`;
+  if (command.meta && command.meta.creationTime === undefined) {
+    command.meta.creationTime = Math.floor(dateInMs / 1000);
+  }
+  return {
+    command,
+    stringify: () => JSON.stringify(command),
+  };
+};
+
+type CAP = (name: string, ...args: any[]) => ICapabilityItem;
+
+type ExtractType<T> = T extends (cmd: { payload: infer A }) => any
+  ? A extends { funs: infer F }
+    ? F extends Array<infer I>
+      ? UnionToIntersection<I> extends { capability: infer C }
+        ? C
+        : CAP
+      : CAP
+    : CAP
+  : CAP;
+
+interface ISinger {
+  (
+    first:
+      | string
+      | { pubKey: string; scheme?: 'ED25519' | 'ETH'; address?: string },
+  ): () => Pick<ICommand, 'signers'>;
+  <T extends any>(
+    first:
+      | string
+      | { pubKey: string; scheme?: 'ED25519' | 'ETH'; address?: string },
+    capability: (withCapability: ExtractType<T>) => ICapabilityItem[],
+  ): T;
+}
+
+export const signer: ISinger = ((
+  first:
+    | string
+    | { pubKey: string; scheme?: 'ED25519' | 'ETH'; address?: string },
+  capability: (
+    withCapability: (
+      name: string,
+      ...args: any[]
+    ) => { name: string; args: any[] },
+  ) => ICapabilityItem[],
+): any => {
+  const {
+    pubKey,
+    scheme = 'ED25519',
+    address = undefined,
+  } = typeof first === 'object' ? first : { pubKey: first };
+  let clist: undefined | Array<{ name: string; args: any[] }>;
+  if (typeof capability === 'function') {
+    clist = capability(((name: string, ...args: any[]) => ({
+      name,
+      args,
+    })) as any);
+  }
+  return () => ({
+    signers: [
+      {
+        ...prop('pubKey', pubKey),
+        ...prop('scheme', scheme),
+        ...prop('address', address),
+        ...prop('clist', clist),
+      },
+    ],
+  });
+}) as any;
```

Comment:
I suppose the `any` is just a WIP?

Reply:
this function infers input types from return type 🙃
I have split its interface from the implementation.  check `ISigner`
its something like this

```typeScript
const myFn = <T extends any>(cb: (arg: T) => string): T =>
  ((arg: any) => arg) as any;

const builder = (arg: { name: string; args: any[] }) => {};

// myFn argument has the same type as builder argument
builder(myFn((arg: { name: string; args: any[] }) => arg.name));
```